Source/JavaScriptCore/runtime/AtomicsObject.cpp

```diff
@@ -442,28 +445,60 @@ JSValue atomicsWaitImpl(JSGlobalObject* globalObject, JSArrayType* typedArray, u
         timeout = std::max(Seconds::fromMilliseconds(timeoutInMilliseconds), 0_s);
 
     if (!vm.m_typedArrayController->isAtomicsWaitAllowedOnCurrentThread()) {
-        throwTypeError(globalObject, scope, "Atomics.wait cannot be called from the current thread."_s);
+        throwTypeError(globalObject, scope, makeString("Atomics."_s, (isAsync ? "waitAsync"_s : "wait"_s), " cannot be called from the current thread."_s));
         return { };
     }
 
+    JSValue result;
     bool didPassValidation = false;
-    ParkingLot::ParkResult result;
-    {
-        ReleaseHeapAccessScope releaseHeapAccessScope(vm.heap);
-        result = ParkingLot::parkConditionally(
-            ptr,
-            [&] () -> bool {
-                didPassValidation = WTF::atomicLoad(ptr) == expectedValue;
-                return didPassValidation;
-            },
-            [] () { },
-            MonotonicTime::now() + timeout);
+    if (isAsync) {
+        JSObject* obj = constructEmptyObject(getVM(globalObject), globalObject->objectStructureForObjectConstructor());
+
+        JSValue async;
+        JSValue value;
+        didPassValidation = WTF::atomicLoad(ptr) == expectedValue;
+        if (!didPassValidation) {
+            async = jsBoolean(false);
+            value = vm.smallStrings.notEqualString();
+        } else if (!timeout) {
+            async = jsBoolean(false);
+            value = vm.smallStrings.timedOutString();
+        } else {
+            async = jsBoolean(true);
+            JSPromise* promise = JSPromise::create(vm, globalObject->promiseStructure());
+            WaiterListsManager::singleton().addWaiter(ptr, promise);
+            value = JSValue(promise);
+            RunLoop::current().dispatchAfter(timeout, [ptr, promise] {
+                WaiterListsManager::singleton().timeoutAsyncWaiter(ptr, &promise->globalObject()->vm());
+            });
```

Comment:
Before we send the lambda to the run loop, `WaiterListManager` will `addWaiter` for that `promise`. When we create a`Waiter`, we will append a deferred work with a `ticket` to the `vm` to keep it alive until the `ticket` is consumed by `notifyWaiterImpl`.
```
    Waiter(JSPromise* promise)
        : m_isAsync(true)
    {
        VM& vm = promise->globalObject()->vm();
        vm.apiLock().lock();
        data.ticket = vm.deferredWorkTimer->addPendingWork(vm, promise, { });
        vm.apiLock().unlock();
    }
``` 

Where `notifyWaiterImpl` can be only triggered by two cases:

1. The `timeout` function invoked by the callback function passed to the run loop of the `vm` (this is done by `addWaiter` function).
```
        RunLoop::current().dispatchAfter(timeout, [ptr, waiter] { // <-- this waiter is only used as an ID for finding the actual waiter in the queue.
            WaiterListManager::singleton().timeout(ptr, waiter);
        });
```

2. Some other threads call `notifyWaiter` and then `notifyWaiterImpl ` will check whether the `vm` is still alive by
```
        if (!waiter || !m_registeredVMs.contains(waiter->getVM()))
            return false;
```
This is safe even when the `vm` is destructed since we `unregister` VMs in their destructors.